pkg/k8s/k8s.go

```diff
@@ -0,0 +1,171 @@
+package k8s
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/loft-sh/log/scanner"
+	"github.com/loft-sh/vcluster/pkg/util/loghelper"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+)
+
+const apiserverCmd = "APISERVER_COMMAND"
+const schedulerCmd = "SCHEDULER_COMMAND"
+const controllerCmd = "CONTROLLER_COMMAND"
+
+type command struct {
+	Command []string `json:"command,omitempty"`
+}
+
+func retryAlways(error) bool {
+	return true
+}
+
+func StartK8S(ctx context.Context, apiUp chan struct{}) error {
+	// we need to retry the functions because etcd is started after the syncer, so
+	// the apiservers always fail when we start until etcd is up and running
+	// the controller needs the apiserver service to respond to start successfully
+	// and the service can only be reachable once the syncer can reach the api servers
+
+	backoff := wait.Backoff{Steps: 3, Duration: 2 * time.Second, Factor: 2.0, Jitter: 0}
+
+	eg := &errgroup.Group{}
+
+	apiCommand := &command{}
+	apiEnv, ok := os.LookupEnv(apiserverCmd)
+	if ok {
+		err := yaml.Unmarshal([]byte(apiEnv), apiCommand)
+		if err != nil {
+			return fmt.Errorf("parsing apiserver command %s: %w", apiEnv, err)
+		}
+		eg.Go(func() error {
+			return retry.OnError(backoff, retryAlways, func() error {
+				return RunCommand(ctx, *apiCommand, "apiserver")
+			})
+		})
+	}
+
+	isUp := waitForAPI(ctx)
+	if !isUp {
+		return errors.New("waited until timeout for the api to be up, but it never did")
+	}
+	close(apiUp)
+
+	controllerCommand := &command{}
+	controllerEnv, ok := os.LookupEnv(controllerCmd)
+	if ok {
+		err := yaml.Unmarshal([]byte(controllerEnv), controllerCommand)
+		if err != nil {
+			return fmt.Errorf("parsing controller command %s: %w", controllerEnv, err)
+		}
+		eg.Go(func() error {
+			return retry.OnError(backoff, retryAlways, func() error {
```

Comment:
Why retry here? I think we should fatal if the command fatals

Reply:
yeah, I forgot to remove it, because before I was not waiting for the api to be up and it would always crash